### PR TITLE
Add recording of read assets to InMemoryAssetReader

### PR DESCRIPTION
**Pull request by nshahan**
Gives ability to verify which assets have been read by a Builder.

### CODE REVIEW

File: build_test/lib/src/in_memory_reader.dart [hunk 1]
```diff
@@ -9,29 +9,39 @@ import 'package:glob/glob.dart';
 
 import 'in_memory_writer.dart';
 
+/// An [AssetReader] that records which assets have been read to [readAssets].
+abstract class RecordingAssetReader implements AssetReader {
+  Iterable<AssetId> get readAssets;
```
Comment (jakemac53) [inline build_test/lib/src/in_memory_reader.dart]: nit: `assetsRead`? `readAssets` sounds like an action :)

File: build_test/lib/src/in_memory_reader.dart [hunk 2]
```diff
@@ -9,29 +9,39 @@ import 'package:glob/glob.dart';
...
   /// Create a new asset reader that contains [sourceAssets].
   ///
   /// May optionally define a [rootPackage], which is required for some APIs.
   InMemoryAssetReader({Map<AssetId, DatedValue> sourceAssets, this.rootPackage})
-      : assets = sourceAssets ?? <AssetId, DatedValue>{};
+      : assets = sourceAssets ?? <AssetId, DatedValue>{},
+        readAssets = new Set<AssetId>();
 
   @override
   Future<bool> canRead(AssetId id) async => assets.containsKey(id);
```
Comment (jakemac53) [inline build_test/lib/src/in_memory_reader.dart]: nit: we should probably add these assets as well, calling `canRead` on an asset adds a dependency on it just like `readAs*`

File: build_test/lib/src/in_memory_reader.dart (hunk 1 above)
Comment (nshahan) [reply build_test/lib/src/in_memory_reader.dart]: Yeah English is tough. It's hard to distinguish read (present tense) vs read (past tense). I'll change it.

File: build_test/lib/src/in_memory_reader.dart (hunk 2 above)
Comment (nshahan) [reply build_test/lib/src/in_memory_reader.dart]: Done.

Comment (nshahan) [reply build_test/lib/src/in_memory_reader.dart]: Just to clarify, does this mean that you would expect there to be a dependency on a file that returns false from canRead? In an offline conversation, I gathered that this is true.

**Pull request merged**